src/api/accounts.rs

```diff
@@ -0,0 +1,515 @@
+        let v = signature.v + 27;
+
+        let signature_bytes = Bytes({
+            let mut bytes = Vec::with_capacity(65);
+            bytes.extend_from_slice(&signature.r[..]);
+            bytes.extend_from_slice(&signature.s[..]);
+            bytes.push(v);
+            bytes
+        });
+
+        Ok(SignedData {
+            message,
+            message_hash: message_hash.into(),
+            v,
+            r: signature.r.into(),
+            s: signature.s.into(),
+            signature: signature_bytes,
+        })
+    }
+
+    /// Recovers the Ethereum address which was used to sign the given data.
+    pub fn recover<S, Sig>(&self, message: S, signature: Sig) -> Result<Address, Error>
+    where
+        S: AsRef<str>,
+        Sig: IntoSignature,
+    {
+        let message_hash = self.hash_message(message);
+        let signature = signature.into_signature();
+
+        let public_key = signature.recover(&message_hash[..]).map_err(EthsignError::from)?;
+
+        Ok(public_key.address().into())
+    }
+}
+
+type MaybeReady<T, R> = Either<FutureResult<R, Error>, CallFuture<R, <T as Transport>::Out>>;
+
+type TxParams<T> = Join3<MaybeReady<T, U256>, MaybeReady<T, U256>, MaybeReady<T, String>>;
+
+/// Future resolving when transaction signing is complete
+pub struct SignTransactionFuture<T: Transport> {
+    tx: TransactionParameters,
+    key: SecretKey,
+    inner: TxParams<T>,
+}
+
+impl<T: Transport> SignTransactionFuture<T> {
+    /// Creates a new SignTransactionFuture with accounts and transaction data.
+    pub fn new(accounts: &Accounts<T>, tx: TransactionParameters, key: &SecretKey) -> SignTransactionFuture<T> {
+        macro_rules! maybe {
+            ($o: expr, $f: expr) => {
+                match $o.clone() {
+                    Some(value) => Either::A(future::ok(value)),
+                    None => Either::B($f),
+                }
+            };
+        }
+
+        let from = key.public().address().into();
+        let inner = Future::join3(
+            maybe!(tx.nonce, accounts.web3().eth().transaction_count(from, None)),
+            maybe!(tx.gas_price, accounts.web3().eth().gas_price()),
+            maybe!(None, accounts.web3().net().version()),
+        );
+
+        SignTransactionFuture {
+            tx,
+            key: key.clone(),
+            inner,
+        }
+    }
+}
+
+impl<T: Transport> Future for SignTransactionFuture<T> {
+    type Item = SignedTransaction;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let (nonce, gas_price, chain_id) = try_ready!(self.inner.poll());
+
+        let chain_id = chain_id.parse::<u64>().map_err(|e| Error::Decoder(e.to_string()))?;
+        let tx = RawTxParameters {
+            nonce,
+            to: self.tx.to.unwrap_or_default(),
+            value: self.tx.value.unwrap_or_default(),
+            gas_price,
+            gas: self.tx.gas.unwrap_or(TRANSACTION_BASE_GAS_FEE),
```

Comment:
I chose 21K as this is the base transaction amount and exactly what it uses to send eth from one address to another (with no contracts involved). I think if we want to be extra cleaver, we could estimate the gas price when it is not provided.

Reply:
I'd leave the estimation to the user - we can suggest them to run `estimateGas` before making a call. BTW is it easy to go from the payload required for `estimateGas` and the payload that goes to this function? We should consider adding a `From` implementations to make such workflow easier.

Indeed `21k` is ok for simple transfers, but if only you'll try to send some data, we will need more gas. I'm still in favour of increasing the "default sensible" gas here - the user will get the rest refunded anyway and I think it can prevent a class of simple errors. Alternatively we can make `gas` a required field and force the user to always set it.

Reply:
FWIW, web3js requires gas parameter for `web3.eth.accounts.signTransaction`. I like having a default gas as I find it makes things more ergonomic (so we can implement `Default` for `TransactionParameters`). I think I'll hike it up to `100k` then.

I'll add the `Into` and `From` impls between `TransactionParameters` and `CallRequest` which is used for estimate gas.